### PR TITLE
AsyncTargetWrapper - Flush should start immediately without waiting

### PR DESCRIPTION
**Pull request by snakefoot**
AsyncTargetWrapper has the following changes:

- FlushAsync now schedules an "instant" QueueUserWorkItem, instead of waiting for the time-thread-event. Instead of waiting for all LogEvents to be written, then explicit Flush is called on WrappedTarget right after having written all pending logevents.  (minor breaking change)
   - This means that putting a BufferingTargetWrapper inside "targets async=true" will no longer cause the AsyncTargetWrapper to wait for the BufferingTargetWrapper-flush-delay.
    - This also means that Targets with broken Flush-method (not keeping track of pending logevents) will not get help from the AsyncTargetWrapper during flush.
- CloseTarget now flushes all pending messages (instead of just the first BatchSize). CloseTarget-caller was also blocking before, so not much change here.

BufferingTargetWrapper has the following changes:

- FlushAsync now lets the WrappedTarget decide when it has flushed. Instead of waiting for all LogEvents to be written and then call explicit Flush on WrappedTarget. Flush-caller was also blocked before, so not much change here.
    - This also means that Targets with broken Flush-method (not keeping track of pending logevents) will not get help from the BufferingTargetWrapper during flush.
- CloseTarget now flushes any pending messages (minor breaking change, if not bugfix)
- Do not hold Target.SyncRoot while writing messages on Timer-thread (Performance)

Reasons for these changes:
- Closer alignment of AsyncTargetWrapper and BufferingTargetWrapper
- Refactoring to increase the chance of making an instant flush on shutdown, without blocking the finalizer thread too much so it is terminated without anything written (See also #1846)
- Could consider to change AsyncTargetWrapper.FlushAsync to perform synchrone flush instead of ThreadPool.QueueUserWorkItem, but then it would be a major breaking change, as the Flush-caller would be blocked while executing all the writes. (Would avoid spinning up new threads during shutdown, though this is already done by LogFactory.Flush)

Related issues in this area: #134 and #214

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1853)

### CODE REVIEW

**Comment by 304NotModified** [comment]
Compare without whitespace:

https://github.com/NLog/NLog/pull/1853/files?w=1

**Comment by 304NotModified** [comment]
> CloseTarget now flushes any pending messages (minor breaking change, if not bugfix)

Bugfix IMO

> This also means that Targets with broken Flush-method (not keeping track of pending logevents) will not get help from the AsyncTargetWrapper during flush.

Is this good? 

Could we summarize this PR in "Performance & stability improvement for Async & bufferwrapper"? Could you do a small performance test?

**Comment by snakefoot** [comment]
> Is this good

Well it fixes that putting a BufferingTargetWrapper inside "targets async=true" will no longer cause the AsyncTargetWrapper to wait for the BufferingTargetWrapper-flush-delay.

I guess one could make a new special broken-flush-helper-wrapper, that keep track of last written LogEvent, and delays the flush until it has completed. It will ofcourse give an overhead as it would have to allocate  delegate for every write-operation. This wrapper would also fix this situation:
- Async-Timer writes last LogEvent in queue
- Flush is being called, as no LogEvents in queue, then flush completes immediately, without waiting for the LogEvent in flight.

**Comment by snakefoot** [comment]
Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| PerfLogger       | 10.000.000 |   16 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |     6.232 | 1.604.620 |   0 |  45 |  91 |    8.281 |  113,082 |
| This PR    |     6.026 | 1.659.215 |   0 |  45 |  91 |    8.421 |  115,777 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| PerfLogger       | 10.000.000 |  128 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    17.538 |   570.164 | 248 | 351 | 408 |   24.531 |  126,605 |
| This PR    |    17.101 |   584.735 | 251 | 360 | 412 |   24.218 |  131,281 |


Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| PerfSyncLogger   | 10.000.000 |   16 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |     7.614 | 1.313.202 |   0 |   0 | 489 |    7.562 |   35,496 |
| This PR    |     7.033 | 1.421.758 |   0 |   0 | 492 |    7.265 |   35,387 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| PerfSyncLogger   | 10.000.000 |  128 |       1 |       1 |

| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    15.026 |   665.511 |   0 |   0 | 1240 |   15.218 |   36,055 |
| This PR    |    14.169 |   705.733 |   0 |   0 | 1240 |   14.421 |   35,910 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| Logger           | 10.000.000 |    8 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    15.229 |   656.638 |   0 | 105 | 217 |   15.171 |  109,406 |
| This PR    |    15.464 |   646.649 |   0 | 101 | 211 |   13.703 |  109,711 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| Logger           | 10.000.000 |   16 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    15.574 |   642.077 |   0 | 108 | 222 |   15.421 |  106,434 |
| This PR    |    15.965 |   626.344 |   0 | 100 | 206 |   14.234 |  111,664 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| Logger           | 10.000.000 |  128 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    24.033 |   416.081 | 1024 | 1117 | 1123 |   54.656 |  129,785 |
| This PR    |    24.915 |   401.352 | 1008 | 1094 | 1098 |   53.718 |  123,977 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| Logger           | 10.000.000 |   16 |       2 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    31.092 |   321.617 |   0 | 212 | 441 |   60.109 |  112,711 |
| This PR    |    31.963 |   312.852 |   0 | 204 | 421 |   49.281 |  119,473 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| Logger           |  3.000.000 |   16 |       4 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    18.532 |   161.882 |   0 | 108 | 224 |   66.640 |  119,410 |
| This PR    |    19.682 |   152.416 |   0 | 106 | 218 |   68.750 |  130,020 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| Logger           |  3.000.000 |   16 |       8 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    45.495 |    65.940 |   0 | 132 | 268 |  169.687 |  152,699 |
| This PR    |    45.875 |    65.394 |   0 | 130 | 261 |  170.687 |  153,348 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SyncLogger       | 10.000.000 |    8 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    19.606 |   510.022 |   0 |   0 | 844 |   19.625 |   35,902 |
| This PR    |    19.343 |   516.960 |   0 |   0 | 844 |   19.406 |   35,691 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SyncLogger       | 10.000.000 |   16 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    19.824 |   504.434 |   0 |   0 | 900 |   19.953 |   35,891 |
| This PR    |    19.438 |   514.432 |   0 |   0 | 899 |   19.656 |   35,715 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SyncLogger       | 10.000.000 |  128 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    26.825 |   372.786 |   0 |   1 | 1653 |   26.828 |   36,352 |
| This PR    |    26.362 |   379.323 |   0 |   1 | 1649 |   26.703 |   36,434 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SimpleLogger     | 10.000.000 |    8 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    17.427 |   573.809 |   0 |  69 | 138 |   20.031 |  130,895 |
| This PR    |    17.123 |   583.984 |   0 |  68 | 137 |   19.968 |  131,711 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SimpleLogger     | 10.000.000 |   16 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    18.224 |   548.718 |   0 |  74 | 148 |   21.296 |  130,762 |
| This PR    |    17.581 |   568.781 |   0 |  71 | 143 |   20.093 |  132,430 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SimpleLogger     | 10.000.000 |  128 |       1 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    31.832 |   314.148 | 618 | 804 | 820 |   40.578 |  101,430 |
| This PR    |    30.881 |   323.823 | 618 | 823 | 833 |   40.156 |  100,910 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SimpleLogger     | 10.000.000 |   16 |       2 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    36.860 |   271.293 |   0 | 141 | 282 |   46.968 |  132,020 |
| This PR    |    35.618 |   280.751 |   0 | 141 | 283 |   46.031 |  131,629 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SimpleLogger     |  3.000.000 |   16 |       4 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    25.131 |   119.370 |   0 |  84 | 168 |   36.265 |  134,418 |
| This PR    |    23.668 |   126.751 |   0 |  87 | 175 |   36.015 |  130,324 |

Executing warmup run... (.NET=4.6.1586.0, Platform=32bit)
Executing performance test...

| Logger Name      | Messages   | Size | Threads | Loggers |
|------------------|------------|------|---------|---------|
| SimpleLogger     |  3.000.000 |   16 |       8 |       1 |


| Test Name  | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Mem (MB) |
|------------|-----------|-----------|-----|-----|-----|----------|----------|
| Master    |    57.450 |    52.218 |   0 | 163 | 326 |  117.562 |  135,309 |
| This PR    |    51.917 |    57.783 |   0 | 168 | 337 |  118.828 |  133,051 |

**Comment by 304NotModified** [comment]
Fixes https://github.com/NLog/NLog/issues/1833 ?

File: tests/NLog.UnitTests/Targets/Wrappers/AsyncTargetWrapperTests.cs [hunk 1]
```diff
@@ -484,7 +494,12 @@ protected override void FlushAsync(AsyncContinuation asyncContinuation)
             {
                 Interlocked.Increment(ref this.FlushCount);
                 ThreadPool.QueueUserWorkItem(
-                    s => asyncContinuation(null));
+                    s =>
+                    {
+                        while (0 != this.PendingWriteCount)
+                            Thread.Sleep(1);
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Targets/Wrappers/AsyncTargetWrapperTests.cs]: it this not causing a high CPU? Would wait 10-100ms instead of 1ms

**Comment by snakefoot** [comment]
> Fixes #1833 ?

Yes WriteEventsInQueue() now checks logEvents.Length before it writes.

File: tests/NLog.UnitTests/Targets/Wrappers/AsyncTargetWrapperTests.cs (hunk 1 above)
Comment (snakefoot) [inline tests/NLog.UnitTests/Targets/Wrappers/AsyncTargetWrapperTests.cs]: Now it has been changed to 10 ms

**Comment by 304NotModified** [comment]
Nice improvement!

**Pull request merged**